### ISSUE TITLE
Evaluation Results are not Consistent in Consecutive Evaluations & Sensitivty to Batch Size

### ISSUE THREAD

**Issue by bruceyo** [ShoufaChen/AdaptFormer#12]
Thank you for sharing this wonderful work! Could you help to look into the following two issues:

1. I tested the code on the HMDB51 dataset, the results can be inconsistent for two consecutive evaluations (run the LINE414 of [main_videp.py](https://github.com/ShoufaChen/AdaptFormer/blob/main/main_video.py) `test_stats = evaluate(data_loader_val, model, device)` twice).

2. For fine-tuning with Swin Transformer, I ran the code with a smaller batch size (i.e., 32) on 4 3090 GPUs several times, and the results for tunning the linear layer are around 71+%. Is the large batch size making such a difference from the reported results of 74%?

Thank you very much in advance!

**Comment by ShoufaChen** [comment]
Hi,

Thanks for your interest.

1. Would you please provide the variance of two consecutive evaluations? Did you only test without training?

2. We didn't test with your settings (i.e., batch size 32). Since you reduce the batch size significantly, I think some hyperparameters need to be tuned accordingly.

**Comment by bruceyo** [comment]
> Hi,
> 
> Thanks for your interest.
> 
> 1. Would you please provide the variance of two consecutive evaluations? Did you only test without training?
> 2. We didn't test with your settings (i.e., batch size 32). Since you reduce the batch size significantly, I think some hyperparameters need to be tuned accordingly.

Hi, thank you so much for the prompt reply!
By running Line 414 of [main_video.py](https://github.com/ShoufaChen/AdaptFormer/blob/main/main_video.py) twice:
` 
test_stats = evaluate(data_loader_val, model, device)
print(f"Accuracy of the network on the {len(dataset_val)} test images: {test_stats['acc1']:.1f}%")
test_stats = evaluate(data_loader_val, model, device)
print(f"Accuracy of the network on the {len(dataset_val)} test images: {test_stats['acc1']:.1f}%")
`
The command is :
`CUDA_VISIBLE_DEVICES=0,1 torchrun --standalone --nnodes=1 \
    --nproc_per_node=2 --master_port=22235 \
    main_video.py \
    --num_frames 8 \
    --sampling_rate 2 \
    --model vit_base_patch16_224 \
    --fulltune \
    --finetune pre_trained/videomae_pretrain_vit_b_1600.pth \
    --output_dir output_hmdb51/ssv2_fulltune_lr001 \
    --batch_size 16 --epochs 90 --blr 0.1 --weight_decay 0.0 --dist_eval \
    --data_path /home/xbyu/data/hmdb51 --data_set HMDB51 \
    --ffn_adapt`

It is the evaluation after every training epoch.
The results are as follows:
[10:47:37.398675] number of params (M): 87.46
[10:47:37.398742] base lr: 1.00e-01
[10:47:37.398752] actual lr: 1.25e-02
[10:47:37.398764] accumulate grad iterations: 1
[10:47:37.398774] effective batch size: 32
[10:47:37.410785] SGD (
Parameter Group 0
    dampening: 0
    foreach: None
    lr: 0.0125
    maximize: False
    momentum: 0.9
    nesterov: False
    weight_decay: 0.0

Parameter Group 1
    dampening: 0
    foreach: None
    lr: 0.0125
    maximize: False
    momentum: 0.9
    nesterov: False
    weight_decay: 0.0
)
[10:47:37.410995] criterion = CrossEntropyLoss()
[10:47:37.411016] Start training for 90 epochs
[10:47:37.411911] log_dir: output_hmdb51/ssv2_fulltune_lr001
[10:47:43.980139] Epoch: [0]  [  0/111]  eta: 0:12:08  lr: 0.000000  loss: 3.9742 (3.9742)  time: 6.5669  data: 2.4233  max mem: 14419
[10:48:16.622125] Epoch: [0]  [110/111]  eta: 0:00:00  lr: 0.001239  loss: 3.6616 (3.7747)  time: 0.2906  data: 0.0002  max mem: 15070
[10:48:16.697082] Epoch: [0] Total time: 0:00:39 (0.3539 s / it)
[10:48:16.727437] Averaged stats: lr: 0.001239  loss: 3.6616 (3.7734)
[10:48:18.606554] Test:  [ 0/48]  eta: 0:01:30  loss: 3.3733 (3.3733)  acc1: 6.2500 (6.2500)  acc5: 25.0000 (25.0000)  time: 1.8750  data: 1.7596  max mem: 15070
[10:48:24.036699] Test:  [47/48]  eta: 0:00:00  loss: 3.8163 (3.7544)  acc1: 6.2500 (5.6209)  acc5: 18.7500 (25.4902)  time: 0.1098  data: 0.0002  max mem: 15070
[10:48:24.199429] Test: Total time: 0:00:07 (0.1556 s / it)
[10:48:24.407291] * Acc@1 **5.686** Acc@5 26.275 loss 3.736
[10:48:24.407493] Accuracy of the network on the 1530 test images: 5.7%
[10:48:26.302295] Test:  [ 0/48]  eta: 0:01:30  loss: 3.3019 (3.3019)  acc1: 12.5000 (12.5000)  acc5: 25.0000 (25.0000)  time: 1.8904  data: 1.7270  max mem: 15070
[10:48:31.608453] Test:  [47/48]  eta: 0:00:00  loss: 3.8210 (3.7441)  acc1: 6.2500 (6.0131)  acc5: 18.7500 (25.3595)  time: 0.1077  data: 0.0002  max mem: 15070
[10:48:31.768639] Test: Total time: 0:00:07 (0.1533 s / it)
[10:48:32.304670] * Acc@1 **6.078** Acc@5 25.882 loss 3.729
[10:48:32.304811] Accuracy of the network on the 1530 test images: 6.1%
[10:48:33.307642] log_dir: output_hmdb51/ssv2_fulltune_lr001
[10:48:35.537274] Epoch: [1]  [  0/111]  eta: 0:04:07  lr: 0.001250  loss: 3.7795 (3.7795)  time: 2.2279  data: 1.8256  max mem: 15070
[10:49:08.242003] Epoch: [1]  [110/111]  eta: 0:00:00  lr: 0.002489  loss: 3.7384 (3.6934)  time: 0.2913  data: 0.0002  max mem: 15070
[10:49:08.402901] Epoch: [1] Total time: 0:00:35 (0.3162 s / it)
[10:49:08.404055] Averaged stats: lr: 0.002489  loss: 3.7384 (3.6680)
[10:49:10.402474] Test:  [ 0/48]  eta: 0:01:35  loss: 3.6926 (3.6926)  acc1: 6.2500 (6.2500)  acc5: 31.2500 (31.2500)  time: 1.9940  data: 1.8655  max mem: 15070
[10:49:15.535313] Test:  [47/48]  eta: 0:00:00  loss: 3.8795 (3.8307)  acc1: 6.2500 (6.6667)  acc5: 18.7500 (22.4837)  time: 0.1037  data: 0.0002  max mem: 15070
[10:49:15.715812] Test: Total time: 0:00:07 (0.1523 s / it)
[10:49:15.900173] * Acc@1 **6.863** Acc@5 22.680 loss 3.809
[10:49:15.900305] Accuracy of the network on the 1530 test images: 6.9%
[10:49:17.882602] Test:  [ 0/48]  eta: 0:01:35  loss: 3.7219 (3.7219)  acc1: 6.2500 (6.2500)  acc5: 31.2500 (31.2500)  time: 1.9799  data: 1.8609  max mem: 15070
[10:49:23.221663] Test:  [47/48]  eta: 0:00:00  loss: 3.8276 (3.8137)  acc1: 6.2500 (6.6667)  acc5: 18.7500 (22.6144)  time: 0.1051  data: 0.0002  max mem: 15070
[10:49:23.401468] Test: Total time: 0:00:07 (0.1563 s / it)
[10:49:23.558848] * Acc@1 **6.797** Acc@5 22.353 loss 3.804
[10:49:23.559017] Accuracy of the network on the 1530 test images: 6.8%
[10:49:24.857928] log_dir: output_hmdb51/ssv2_fulltune_lr001
[10:49:31.502839] Epoch: [2]  [  0/111]  eta: 0:12:17  lr: 0.002500  loss: 3.6328 (3.6328)  time: 6.6437  data: 2.1980  max mem: 15070
[10:50:04.318700] Epoch: [2]  [110/111]  eta: 0:00:00  lr: 0.003739  loss: 3.7075 (3.7261)  time: 0.2921  data: 0.0002  max mem: 15070
[10:50:04.458124] Epoch: [2] Total time: 0:00:39 (0.3568 s / it)
[10:50:04.472337] Averaged stats: lr: 0.003739  loss: 3.7075 (3.7096)
[10:50:06.360826] Test:  [ 0/48]  eta: 0:01:30  loss: 3.3011 (3.3011)  acc1: 6.2500 (6.2500)  acc5: 31.2500 (31.2500)  time: 1.8848  data: 1.7711  max mem: 15070
[10:50:11.594191] Test:  [47/48]  eta: 0:00:00  loss: 4.0105 (3.8955)  acc1: 6.2500 (7.1895)  acc5: 25.0000 (24.5752)  time: 0.1021  data: 0.0002  max mem: 15070
[10:50:11.767886] Test: Total time: 0:00:07 (0.1520 s / it)
[10:50:12.171213] * Acc@1 **6.928** Acc@5 22.941 loss 3.880
[10:50:12.171366] Accuracy of the network on the 1530 test images: 6.9%
[10:50:14.515420] Test:  [ 0/48]  eta: 0:01:52  loss: 3.2539 (3.2539)  acc1: 12.5000 (12.5000)  acc5: 31.2500 (31.2500)  time: 2.3417  data: 2.2185  max mem: 15070
[10:50:19.691111] Test:  [47/48]  eta: 0:00:00  loss: 3.9639 (3.9037)  acc1: 6.2500 (7.0588)  acc5: 18.7500 (23.7908)  time: 0.1020  data: 0.0001  max mem: 15070
[10:50:19.861494] Test: Total time: 0:00:07 (0.1602 s / it)
[10:50:19.863369] * Acc@1 **7.124** Acc@5 22.484 loss 3.872
[10:50:19.863666] Accuracy of the network on the 1530 test images: 7.1%

**Comment by ShoufaChen** [comment]
I am sorry I can't fully understand your issue. It seems that your reported accuracy is very low. Did you train your model?

**Comment by bruceyo** [comment]
> I am sorry I can't fully understand your issue. It seems that your reported accuracy is very low. Did you train your model?

I trained the model. 
For each training epoch, I evaluated the model twice and got inconsistent results. I provided the results of the first three training epochs. 
The issue is not the accuracy but the inconsistent results caused by the evaluation method.
Even though I train the model for 90 epochs (leading to higher accuracy), this issue is still there.

You can reproduce the issue by running the L414-415 of the [main_video.py](https://github.com/ShoufaChen/AdaptFormer/blob/main/main_video.py) two times.

**Comment by ShoufaChen** [comment]
I see. For the evaluation between two training epochs, random selection of video frames exists:(https://github.com/ShoufaChen/AdaptFormer/blob/6967d676c1a5e5a11be2e2768a6e5c604bb043ed/datasets/kinetics.py#L286

 The final results are obtained from https://github.com/ShoufaChen/AdaptFormer/blob/main/main_video.py#L436.

**Comment by bruceyo** [comment]
> I see. For the evaluation between two training epochs, random selection of video frames exists:(
> 
> https://github.com/ShoufaChen/AdaptFormer/blob/6967d676c1a5e5a11be2e2768a6e5c604bb043ed/datasets/kinetics.py#L286
> 
> The final results are obtained from https://github.com/ShoufaChen/AdaptFormer/blob/main/main_video.py#L436.

Thank you:)
